nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,200 @@
+  const std::string & global_frame_id,
+  const tf2::Duration & transform_tolerance,
+  const rclcpp::Duration & source_timeout,
+  const bool base_shift_correction)
+: Source(
+    node, source_name, tf_buffer, base_frame_id, global_frame_id,
+    transform_tolerance, source_timeout, base_shift_correction)
+{
+}
+
+PolygonSource::~PolygonSource()
+{
+  data_sub_.reset();
+}
+
+void PolygonSource::configure()
+{
+  Source::configure();
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<geometry_msgs::msg::PolygonInstanceStamped>(
+    source_topic, qos,
+    std::bind(&PolygonSource::dataCallback, this, std::placeholders::_1));
+}
+
+bool PolygonSource::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data)
+{
+  // Ignore data from the source if it is not being published yet or
+  // not published for a long time
+  if (data_.empty()) {
+    return false;
+  }
+
+  // Remove all data
+  data_.erase(
+    std::remove_if(
+      data_.begin(), data_.end(),
+      [this, curr_time](const geometry_msgs::msg::PolygonInstanceStamped & polygon_stamped) {
+        return curr_time - rclcpp::Time(polygon_stamped.header.stamp) > source_timeout_;
+      }), data_.end());
+
+  tf2::Stamped<tf2::Transform> tf_transform;
+  for (const auto & polygon_instance : data_) {
+    if (base_shift_correction_) {
+      // Obtaining the transform to get data from source frame and time where it was received
+      // to the base frame and current time
+      if (
+        !nav2_util::getTransform(
+          polygon_instance.header.frame_id, polygon_instance.header.stamp,
+          base_frame_id_, curr_time, global_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return false;
+      }
+    } else {
+      // Obtaining the transform to get data from source frame to base frame without time shift
+      // considered. Less accurate but much more faster option not dependent on state estimation
+      // frames.
+      if (
+        !nav2_util::getTransform(
+          polygon_instance.header.frame_id, base_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return false;
+      }
+    }
+    geometry_msgs::msg::PolygonStamped poly_out, polygon_stamped;
+    geometry_msgs::msg::TransformStamped tf = tf2::toMsg(tf_transform);
+    polygon_stamped.header = polygon_instance.header;
+    polygon_stamped.polygon = polygon_instance.polygon.polygon;
+    tf2::doTransform(polygon_stamped, poly_out, tf);
+    convertPolygonStampedToPoints(poly_out, data);
+  }
+  return true;
+}
+
+void PolygonSource::convertPolygonStampedToPoints(
+  const geometry_msgs::msg::PolygonStamped & polygon, std::vector<Point> & data) const
+{
+  /*  Full function generated by GPT */
```

Comment:
I don't know what you want to do about it but just a reminder because I know you feel a certain way about AI-generated code